<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Andreas Handel" />


<title>Some common data processing tasks</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link rel="icon" type="image/png" href="media/MADAlogo.png" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>






<link rel="stylesheet" href="MADAstyle.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MADA Course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    General Information
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./Course_Syllabus.html">Syllabus</a>
    </li>
    <li>
      <a href="./Course_Schedule.html">Schedule</a>
    </li>
    <li>
      <a href="./Course_Communication.html">Communication</a>
    </li>
    <li>
      <a href="./Course_Assessments.html">Assessments</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Modules
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">1. Course and Tools Introduction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Introduction_Course.html">Course Introduction</a>
        </li>
        <li>
          <a href="./Tools_RandRStudio.html">R and Rstudio</a>
        </li>
        <li>
          <a href="./Tools_Github_Introduction.html">GitHub Introduction</a>
        </li>
        <li>
          <a href="./Tools_Github_R_Workflow.html">GitHub and R Workflow</a>
        </li>
        <li>
          <a href="./Tools_Reproducibility.html">RMarkdown &amp; Co</a>
        </li>
        <li>
          <a href="./Tools_Reference_Management.html">Reference Management</a>
        </li>
        <li>
          <a href="./Assessment_Course_Tools_Introduction.html">Assessment: Course and Tools Introduction</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">2. Data Analysis Introduction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Data_Analysis_Motivation.html">Motivating Examples</a>
        </li>
        <li>
          <a href="./Data_Analysis_Overview.html">Data Analysis Overview</a>
        </li>
        <li>
          <a href="./Data_Analysis_Structure.html">Data Analysis Structure</a>
        </li>
        <li>
          <a href="./Assessment_Data_Analysis.html">Assessment: Data Analysis Introduction</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">3. R Coding Basics</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Rcoding_Basics.html">R Coding Basics</a>
        </li>
        <li>
          <a href="./Rcoding_Tidyverse.html">Tidyverse and Friends</a>
        </li>
        <li>
          <a href="./Assessment_Coding.html">Assessment: R Coding Basics</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">4. Contemplating, finding and wrangling data</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Data_Types.html">Data Types</a>
        </li>
        <li>
          <a href="./Data_Finding.html">Finding Data</a>
        </li>
        <li>
          <a href="./Data_Wrangling.html">Data wrangling</a>
        </li>
        <li>
          <a href="./Assessment_Data_Finding_Wrangling.html">Assessment: Data Finding and Wrangling</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">5. Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Visualization_Overview.html">Visualization Overview</a>
        </li>
        <li>
          <a href="./Visualization_in_R.html">Visualization in R</a>
        </li>
        <li>
          <a href="./Assessment_Visualization.html">Assessment: Visualization</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">6. Data Exploration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Data_Exploration.html">Exploring data</a>
        </li>
        <li>
          <a href="./Assessment_Exploration.html">Assessment: Exploration</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">7. Statistical analysis overview</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Statistical_Analysis_Introduction.html">Statistical analysis introduction</a>
        </li>
        <li>
          <a href="./Statistical_Terminology.html">Some Statistical Model Fitting Terminology</a>
        </li>
        <li>
          <a href="./Assessment_Statistical_Analysis.html">Assessment: Statistical Analysis Overview</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">8. Fitting basic statistical models</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Basic_Statistical_Models.html">Basic statistical models</a>
        </li>
        <li>
          <a href="./Model_Fitting_R.html">Fitting statistical models in R</a>
        </li>
        <li>
          <a href="./Assessment_Basic_Models.html">Assessment: Fitting Basic Models</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">9. Evaluating your model fits</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Model_Evaluation_Overview.html">Model evaluation overview</a>
        </li>
        <li>
          <a href="./Model_Evaluation_Performance.html">Model performance evaluation</a>
        </li>
        <li>
          <a href="./Model_Evaluation_Overfitting.html">Model performance &amp; overfitting</a>
        </li>
        <li>
          <a href="./Model_Evaluation_Diagnostics.html">Model Diagostics</a>
        </li>
        <li>
          <a href="./Model_Evaluation_Uncertainty.html">Model Uncertainty</a>
        </li>
        <li>
          <a href="./Assessment_Model_Evaluation.html">Assessment: Model evaluation</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">10. Improving your model fits</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Model_Improvement_Overview.html">Improving model fits overview</a>
        </li>
        <li>
          <a href="./Model_Improvement_Data_Processing.html">Pre-processing data</a>
        </li>
        <li>
          <a href="./Model_Improvement_Subset_Selection.html">Subset selection</a>
        </li>
        <li>
          <a href="./Model_Improvement_Regularization.html">Regularization</a>
        </li>
        <li>
          <a href="./Model_Improvement_Tuning.html">Model tuning</a>
        </li>
        <li>
          <a href="./Assessment_Model_Improvement.html">Assessment: Model Improvement</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">11. Machine Learning Models I</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="./Polynomial_Spline_Models.html">Polynomial and Spline Models</a>
        </li>
        <li>
          <a href="./CART_Models.html">Classification and Regression Trees (CART)</a>
        </li>
        <li>
          <a href="./Many_Tree_Models.html">Many-tree models</a>
        </li>
        <li class="dropdown-header">Assessment: ML models 1</li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Project
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./Project_Information.html">Project information</a>
    </li>
    <li>
      <a href="./Project_Rubric.html">Project rubric</a>
    </li>
    <li>
      <a href="./Project_Review.html">Project review</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Resources
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./Course_Resources.html">Course Resources</a>
    </li>
    <li>
      <a href="./General_Resources.html">General Resources</a>
    </li>
    <li>
      <a href="./Course_Glossary.html">Glossary</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/andreashandel/MADAcourse">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Some common data processing tasks</h1>
<h4 class="author">Andreas Handel</h4>
<h4 class="date">2021-10-23 09:26:06</h4>

</div>


<div id="overview" class="section level1">
<h1>Overview</h1>
<p>This unit provides a quick description of some common data processing tasks that one often needs to do before starting the actual fitting part.</p>
</div>
<div id="learning-objectives" class="section level1">
<h1>Learning Objectives</h1>
<ul>
<li>Become familiar with and able to apply common data processing tasks</li>
<li>Understand the advantages and disadvantages of specific processing tasks</li>
</ul>
</div>
<div id="introductory-comments" class="section level1">
<h1>Introductory comments</h1>
<p>We already discussed tasks that one needs to do to get the data in shape for fitting. We called this data <em>wrangling</em> or <em>cleaning</em>. The tasks described below haven’t shown up yet. These are usually mentioned in the various sources we have been using, but not presented in one place like I do here. Also, the terminology is not quite settled. I call this <em>(pre-)processing data</em>. This is arguably fuzzy since there is a lot of that going on in other stages of the cleaning/wrangling process. In R4DS, they use the term <em>transforming</em> but that term doesn’t quite cover all the topics I’m covering here. Some of these tasks are also discussed and part of the <code>tidymodels</code> framework. Those tasks are usually done closer to the analysis portion of the project, but there is no real hard differentiation. It is best to always keep all of those possible tasks in mind and use them as needed.</p>
</div>
<div id="motivation" class="section level1">
<h1>Motivation</h1>
<p>Beyond general data cleaning, which is always needed, performing specific processing tasks on the data can help for certain types of statistical models. For instance, some methods perform poorly if the data contains predictors that are essentially “noise”, i.e., that are not correlated with the outcome. Predictors that are too correlated with each other can also be a problem for some (but not all) methods. Often, it is useful to have predictors on the same scales or transform outcomes and predictors such that they follow a distribution more suitable for modeling (often a normal distribution). The amount and type of processing needed depends not only on the data itself but also on the statistical method and algorithm you plan to use.</p>
</div>
<div id="common-processing-tasks" class="section level1">
<h1>Common processing tasks</h1>
<div id="variablepredictor-removal" class="section level2">
<h2>Variable/Predictor Removal</h2>
<p>It is often the case that we have more data than we need to answer a scientific question. For almost any analysis, one therefore needs to remove some variables before starting the statistical model fitting process. A simple (and somewhat stupid) example is an identifier (ID) variable for subjects. This contains no meaningful information for the modeling process and should thus be removed before fitting. Other examples are instances where the data was collected for some purpose other than our analysis. In this case, it is likely that there are variables in the data which are irrelevant for our analysis.</p>
<p>Such removal of variables is done on scientific grounds, based on expert opinion. Ideally, you should report in enough detail which parts of the data you included and excluded in that way to allow the reader to make an informed decision if they agree with what you did. And of course, you should – as much as possible – also provide the raw data and the R scripts which do and document removal of specific variables – such that someone who doesn’t agree with your choice could re-run the analysis with different inclusion/exclusion criteria.</p>
<p>Sometimes, you might have variables that could, in principle, be useful, but the reported values show little diversity and thus contain little information. For instance, if you had a sample of several hundred individuals and only 3 of them were smokers, then it might not be useful to include the smoking variable for the analysis of this dataset, even if, in general, it might be worth considering. Such variables that do not contain much information are called “near-zero variance” variables. Some models perform better if those variables are removed. Other modeling approaches do not care since they have built-in mechanisms to remove variables that are not useful in predicting the outcome.</p>
<p>Another instance where removing variables might be useful is if predictors are strongly collinear/correlated. A trivial example is if you have height reported as both inches and centimeters in your data. Obviously, one of them should be removed. Other variables might not be so obviously containing the same information, but might be related enough (collinear) that including both makes the model performance worse. An example might be age and height among children.</p>
<p>Another example we already disused are variables with missing values. In this case, you might want to remove the variable, or the observations, or a mix of them.</p>
<p>In R, removing variables or observations is easily done with the <code>select</code> and <code>filter</code> functions from <code>dplyr</code>. There are a variety of good packages for dealing with missing/NA values. The <a href="https://cran.r-project.org/web/packages/naniar/index.html"><code>naniar</code></a> package is one of them.</p>
<p><strong>All of these inclusion/exclusion decisions are based on expert (that’s you!) judgment. There is no magic recipe you can follow. Some statistical methods can help (see below), but they are only so useful. If someone on the internet tries to sell a magical way of knowing how to process your data before fitting, run away!</strong></p>
</div>
<div id="variable-transformation" class="section level2">
<h2>Variable transformation</h2>
<p>It is often a good idea to transform variables. For instance if you have a model with continuous and categorical predictors, and you want to compare the coefficients returned from some model, you need to have the predictors on the same scales, which you can accomplish by <strong>standardizing</strong> predictors through <strong>centering</strong> (transforming the data so the mean is a specific value) and <strong>scaling</strong> (transforming the data so that it falls within a specific range). The most common method of centering is subtracting the mean, and the most common method of scaling is dividing by the standard deviation. The drawback of such a transformation is that the model results might be harder to interpret and might have to be ‘back-transformed’ to have biological meaning.</p>
<p>Other common used transformations are converting to log units, since often the log scale is the biologically relevant one. This also reduces the impact of very high and very low values.</p>
<p>It is also often the case that some numerical methods your statistical model uses work better when you transform variables. It is always good to check the manual for any statistical model (and its numerical implementation) to see what is described and recommended there.</p>
<p>Whenever you do some kind of normalization or transformation, check the result of such a transformation to ensure nothing went wrong (e.g., you didn’t accidentally divide by 0 or take the log of a negative number).</p>
<p>There are many ways to do variable transformation in R. In this course we will focus for our model fitting approaches on the <a href="https://www.tidymodels.org/">tidymodels</a> set of packages. Within that set of packages, the <a href="https://recipes.tidymodels.org/">recipes package</a> is used for variable transformation.</p>
</div>
<div id="feature-engineering-variable-creation" class="section level2">
<h2>Feature engineering (variable creation)</h2>
<p>The term <em>feature engineering</em> is often used in machine learning. It basically means creating new variables from the existing ones. The hope is that the new variables are better at describing/predicting/correlating with the outcome. (<strong>Feature</strong> is another word used for <strong>predictor</strong> or <strong>independent</strong> variable. The term <strong>feature</strong> is especially common in the machine learning literature). For instance, if we want to assess the impact of weight on diabetes, using weight itself might not be the best option since it is confounded by height. A 200lb person who’s 5 feet tall is likely at greater risk than someone who is 6 feet tall. We can instead use weight and height and compute BMI and then include that variable/feature in our model instead. In the ML terminology, we just <em>engineered the feature of BMI from the features weight and height</em>. Or in more everyday wording <em>we created a new variable BMI from the two existing variables weight and height</em>.</p>
<p>Creating the right features that are most meaningful for a model needs to be guided by expert knowledge, and often differentiates the models with good predictive performance from those that are not so good. There are whole books focused on feature engineering (e.g. <a href="http://www.feat.engineering/">Feature Engineering and Selection: A Practical Approach for Predictive Models</a> ). The <a href="https://recipes.tidymodels.org/">recipes package</a> mentioned above helps with some feature engineering tasks. We won’t use it too much in the course exercises, but keep in mind that it is always an option to create new variables from your existing ones. (One could think of variable transformation described in the previous section as a very simple form of feature engineering, i.e., we create a new variable/feature which is basically just the old one but somehow transformed.)</p>
</div>
<div id="feature-reduction-variable-reduction" class="section level2">
<h2>Feature Reduction (variable reduction)</h2>
<p>This could be considered one type of feature engineering. In some areas of science, this is a step that’s almost always required. For instance, with <strong>-omics</strong> type data (proteomics, transcriptomics, metabolomics, etc.), it is not uncommon to measure 1000s of variables (e.g., gene expression levels) for one individual person. Thus those datasets often have many more predictors than observations (individuals/samples in the study). This means some models might not work at all (e.g. a standard linear or logistic model) and other models might work but take way too long to run.</p>
<p>It is therefore often necessary to reduce the number of predictors. Manual removal of predictors based on biological/expert knowledge, as described above, is one option. But with thousands of predictors, and often no clear <em>a priori</em> idea of which ones are biologically meaningful, this quickly becomes unfeasible. Another option is to use a statistical approach with the goal to find a set of new predictors of size <span class="math inline">\(m\)</span>, made up of combinations of the <span class="math inline">\(p\)</span> old ones, such that <span class="math inline">\(m \ll p\)</span> (<span class="math inline">\(m\)</span> is much less than <span class="math inline">\(p\)</span>). An approach called Principal Component Analysis (PCA) can be used to find such a smaller set of new predictors that capture most of the information contained in the larger set. One drawback of PCA is that it ignores the relationship between predictors and outcome. That means it tries to reduce the dimension/number of predictors by looking for correlations among them, but without paying attention to the potential correlations between predictors and outcome. This can at times lead to sub-optimal performance of the model. To keep predictors that are most associated with the outcome in the model, one can use a method called Partial Least Squares (PLS). Other approaches exist. The overall problem is that the new set of predictors is harder to interpret, and thus insights gained from the model are somewhat reduced. We won’t have time to look in much detail at any of those approaches, but the resources we have been using for this class cover some of the approaches, so check there if interested.</p>
</div>
</div>
<div id="processing-one-should-avoid" class="section level1">
<h1>Processing one should avoid</h1>
<p>There are certain very common ways data gets processed that should, in fact, be avoided. Here are some things <strong>you should not do!</strong></p>
<p><strong>Discretizing/binning continuous variables.</strong> Doing so leads to loss of power and the binning is almost always arbitrary. E.g., there is no reason to stick people into age categories - keep age continuous (no matter how many Epi papers you see where they do that kind of stuff.) Unfortunately, binning/categorizing of continuous variables is very commonly done, but almost never is there a good reason to do so, other than custom and people not being familiar with other approaches (exceedingly bad reasons for doing things). In the statistical literature, this unhealthy urge to dichotomize/discretize has been labeled <strong>Dichotomania</strong>. <a href="https://www.fharrell.com/post/errmed/#catg">This reference and links therein</a> discusses dichotomania related to chopping up of variables. Another dichotomania exists regarding misinterpretation of statistical results and especially the “magical” 0.05 <em>p</em>-value threshold, which is thorough abused throughout the scientific literature. For some discussion on this <a href="https://academic.oup.com/aje/article/186/6/639/3886035">see e.g., here</a> or <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/sim.2331">this review</a>.</p>
<p>Sometimes, the reason given for binning is that there might not be a linear relation between predictor and outcome, e.g., between age and risk of some disease. That is certainly possible. But the right approach for this is almost always to change the model and use a more flexible statistical model that allows a relation between predictor and outcome that is not linear.</p>
<p>There are occasions when discretizing is warranted. For instance at the end of an analysis, one might need to have a decision rule that determines further action (e.g., treat or don’t treat with statin drugs) based on some continuous quantity (e.g., cholesterol level). However, the analysis should still be done with continuous variables and any binning only done at the end.</p>
<p><strong>Variable transformation</strong> is – as I mentioned above – sometimes a good idea. However, it should not be used without a reason, and sometimes there are better alternatives. For instance if your outcome is not normally distributed, then maybe use a model that more closely mimics your data. For instance, if your data come from a process that likely produces Poisson distributed data, you can use a generalized linear model with a Poisson distribution assumption. Similarly, you do not always need to transform predictors: at times transformations make interpretability harder and it might be better to not transform.</p>
</div>
<div id="general-suggestions" class="section level1">
<h1>General suggestions</h1>
<p>Once you have an idea of the kinds of models you want to apply, figure out if they require or could benefit from some of the data pre-processing steps just described. If yes, implement those processing steps. If a specific algorithm allows it, you can always try to fit the data with and without processing. The differences (or lack of differences) might tell you something useful. At times, this might be a trial-and-error process, where you first do not pre-process, realize the model doesn’t work, then figure out what further processing is needed, then try again.</p>
</div>
<div id="further-resources" class="section level1">
<h1>Further Resources</h1>
<p>For some more reading see <a href="https://bradleyboehmke.github.io/HOML/process.html">Chapter 3 of HMLR</a>. The above mentioned book <a href="http://www.feat.engineering/">Feature Engineering and Selection: A Practical Approach for Predictive Models</a> is also a good resource. The <a href="https://www.tidymodels.org/">tidymodels website</a> also has articles and tutorials that cover some of these topics.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
